[Validator] Modernize Custom constraints article (finishes #13898) (#17335)
symfony/symfony-docs: validation/custom_constraint.rst

~~~~~~~~~~~~~~~~~~~~~~~~~~

Besides validating a single property, a constraint can have an entire class
as its scope.

For instance, imagine you also have a ``PaymentReceipt`` entity and you
need to make sure the email of the receipt payload matches the user's
email. First, create a constraint and override the ``getTargets()`` method::

    // src/Validator/ConfirmedPaymentReceipt.php
    namespace App\Validator;

    use Symfony\Component\Validator\Constraint;

    /**
     * @Annotation
     */
    class ConfirmedPaymentReceipt extends Constraint
    {

>         public string $userDoesNotMatchMessage = 'User\'s e-mailaddress does not match that of the receipt';

Reviewer: ```suggestion
        public string $userDoesNotMatchMessage = 'User\'s e-mail address does not match that of the receipt';
```

        public function getTargets(): string
        {
            return self::CLASS_CONSTRAINT;
        }
    }

> Now, the constraint validator will get an object as first argument to

Reviewer: ```suggestion
Now, the constraint validator will get an object as the first argument to
```

    associated. Use any :doc:`valid PropertyAccess syntax </components/property_access>`
    to define that property.

A class constraint validator must be applied to the class itself:

.. configuration-block::

    .. code-block:: php-annotations

        // src/Entity/PaymentReceipt.php
        namespace App\Entity;

        use App\Validator\ConfirmedPaymentReceipt;

        /**
         * @ConfirmedPaymentReceipt
         */
        class PaymentReceipt
        {
            // ...
        }

    .. code-block:: yaml

        # config/validator/validation.yaml

>         AppBundle\Entity\PaymentReceipt:

Reviewer: ```suggestion
        App\Entity\PaymentReceipt:
```
